### PR TITLE
oslogin: compat.h to suport FreeBSD

### PR DESCRIPTION
**Pull request by helen-fornazier**
Add compat.h to define different code between Linux and FreeBSD systems

### CODE REVIEW

**Comment by illfelder** [comment]
Hi Helen, the change looks good but we still want to verify it doesn't break non-BSD builds. I'll merge when Danny performs the validation and gives the go-ahead.

File: google_compute_engine_oslogin/nss_module/nss_oslogin.cc [hunk 1]
```diff
@@ -103,4 +104,28 @@ int _nss_oslogin_getpwnam_r(const char *name, struct passwd *result,
 // nss_getpwent_r() is intentionally left unimplemented. This functionality is
 // now covered by the nss_cache binary and nss_cache module.
 
+void _nss_oslogin_getpwent_r() {}
+void _nss_oslogin_endpwent() {}
+void _nss_oslogin_setpwent() {}
```
Comment (Sarsate) [inline google_compute_engine_oslogin/nss_module/nss_oslogin.cc]: When I tested this on debian-9, getpwent_r needs to return a valid NSS_STATUS. Without it, every call will complain about an "illegal status in __nss_nextAborted".

In our case it's actually fine to leave these out completely, but if BSD needs these to be explicitly defined, then getpwent_r() should return NSS_STATUS_NOTFOUND, and the (end|set)pwent() methods should return NSS_STATUS_SUCCESS.

**Comment by Sarsate** [comment]
Looks good! Other than the one comment, everything else builds and works as expected.

**Comment by illfelder** [comment]
The CLA was countersigned this morning. Hopefully it's fixed now.

**Comment by illfelder** [comment]
It looks like the CLA went live yesterday afternoon. Hopefully this will re-trigger the CLA check.

**Comment by illfelder** [comment]
The CLA is fixed, but I don't know how to trigger the check on your pull requests. Can you try replying with "I signed it!" or making any update to your PR that may trigger the presubmits to run again?

**Comment by lucaskanashiro** [comment]
I signed it!

**Comment by helen-fornazier** [comment]
> So there's good news and bad news.
> 
>  The good news is that everyone that needs to sign a CLA (the pull request submitter and all commit authors) have done so. Everything is all good there.
> 
>  The bad news is that it appears that one or more commits were authored or co-authored by someone other than the pull request submitter. We need to confirm that all authors are ok with their commits being contributed to this project. Please have them confirm that here in the pull request.

I confirm

> 
> _Note to project maintainer: This is a terminal state, meaning the `cla/google` commit status will not change from this state. It's up to you to confirm consent of all the commit author(s), set the `cla` label to `yes` (if enabled on your project), and then merge this pull request when appropriate._

**Comment by lucaskanashiro** [comment]
@illfelder as you requested, we have tested these changes using `oslogin-ssh` image tests against the following Linux images:

* CentOS 6 and 7
* Debian 9
* Ubuntu 14, 16 and 18 LTS

And all of them worked as expected (the tests successfully passed). So, thumbs up for these modifications.

**Comment by hopkiw** [comment]
Please update the rest of the oslogin PAM modules to use the new code.
Can you also elaborate on the use of NSS_ defines you've added? Feel free to point me to any relevant documentation.

**Comment by lucaskanashiro** [comment]
> Please update the rest of the oslogin PAM modules to use the new code.

Of course. We have some new `pam_syslog` occurrences added by the latest commits, I'll update that and validate the rest of the PAM module modifications in FreeBSD.

> Can you also elaborate on the use of NSS_ defines you've added? Feel free to point me to any relevant documentation.

This is the documentation related to `NSS_*` : https://www.gnu.org/software/libc/manual/html_node/NSS-Modules-Interface.html

**Comment by lucaskanashiro** [comment]
I have updated the patch with the latest modifications and I figured out that modifications made to `google_compute_engine_oslogin/bin/google_oslogin_control` by [this commit](https://github.com/GoogleCloudPlatform/compute-image-packages/commit/6e845ff86d6adf1eb46a238c29d5ab16fbe873fe#diff-787c896d3c64239b68d533b8912d2785) breaks PAM module in FreeBSD.

After install the guest agent package (built using the proposed modifications to make it FreeBSD compatible) and enable OS Login in the instance level, I got the error below when tried to log in:

```
Nov 27 12:34:09 kanashiro-freebsd-12-oslogin-compat-new sshd[1065]
: in openpam_parse_chain(): /etc/pam.d/sshd(21): missing or invalid control flag
Nov 27 12:34:09 kanashiro-freebsd-12-oslogin-compat-new sshd[1065]: in openpam_parse_chain(): /etc/pam.d/sshd(21): 
missing or invalid control flag
Nov 27 12:34:09 kanashiro-freebsd-12-oslogin-compat-new sshd[1065]: fatal: PAM: initialisation failed
Nov 27 12:34:09 kanashiro-freebsd-12-oslogin-compat-new sshd[1065]: fatal: PAM: initialisation failed
```

I did check the `/etc/pam.d/sshd` file and I had this:
```
# Added by Google Compute Engine OS Login.
account    [success=ok ignore=ignore default=die] pam_oslogin_login.so
account    [success=ok default=ignore] pam_oslogin_admin.so
```
Seems that these flags are not supported in FreeBSD. I went through the function `modify_pam_sshd()` in `google_compute_engine_oslogin/bin/google_oslogin_control` that modifies this file and saw this comment: `# TODO: SLES, FreeBSD, others?`. In short, I guess this OS Login control scripts is not FreeBSD compatible again, although [this previous commit](https://github.com/GoogleCloudPlatform/compute-image-packages/commit/29a80fe6ec4901df63a6e8bff6ed1c283a202180) that had made it works before. So, we need to fix this too.

**Comment by illfelder** [comment]
Thanks for checking @lucaskanashiro! Please let me know when the PR is ready for another review.

**Comment by lucaskanashiro** [comment]
This PR is now in sync with latest PAM module modifications. Some other FreeBSD related issues were found and already addressed: https://github.com/GoogleCloudPlatform/compute-image-packages/pull/699 and https://github.com/GoogleCloudPlatform/compute-image-packages/pull/702.

With those modifications and after tweak the instance_config file (when we update the FreeBSD package again we will add this new config), the OS Login image tests are successfully passing in FreeBSD.

**Comment by lucaskanashiro** [comment]
@illfelder @hopkiw Now this PR is up-to-date with the repo layout modification, ready for review.

**Pull request merged**